packages/app/src/specs/flaky-badge/FlakyInformation.vue

```diff
@@ -0,0 +1,107 @@
+<template>
+  <Tooltip
+    v-if="isFlaky"
+    placement="top"
+    :is-interactive="true"
+    class="h-16px"
+    :hide-delay="0"
+    :distance="10"
+    style="width: fit-content"
+  >
+    <ExternalLink
+      :href="dashboardUrl"
+      class="hocus:no-underline"
+    >
+      <FlakyBadge />
+    </ExternalLink>
+    <template #popper="{ shown }">
+      <ExternalLink
+        v-if="shown && projectId && specPath"
+        :href="dashboardUrl"
+        class="hocus:no-underline"
+      >
+        <FlakySpecSummaryAdapter
+          :project-id="projectId"
+          :from-branch="fromBranch"
+          :spec-path="specPath"
+          :spec-name="specName"
+          :spec-extension="specExtension"
+        />
+      </ExternalLink>
+    </template>
+  </Tooltip>
+</template>
+
+<script setup lang="ts">
+
+import ExternalLink from '@cy/gql-components/ExternalLink.vue'
+import type { FlakyInformationProjectFragment, FlakyInformationSpecFragment, FlakyInformationCloudSpecFragment } from '../../generated/graphql'
+import { gql } from '@urql/vue'
+import { computed } from 'vue'
+import Tooltip from '@packages/frontend-shared/src/components/Tooltip.vue'
+import FlakySpecSummaryAdapter from './FlakySpecSummaryAdapter.vue'
+import FlakyBadge from './FlakyBadge.vue'
+import { getUrlWithParams } from '@packages/frontend-shared/src/utils/getUrlWithParams'
+
+gql`
+fragment FlakyInformationProject on CurrentProject {
+  id
+  projectId
+  branch
+}
+`
+
+gql`
+fragment FlakyInformationSpec on Spec {
+  id
+  relative
+  fileName
+  fileExtension
+}
+`
+
+gql`
+fragment FlakyInformationCloudSpec on RemoteFetchableCloudProjectSpecResult {
+  id
+  data {
+    ... on CloudProjectSpec {
+      id
+      isConsideredFlaky(fromBranch: $fromBranch)
+      flakyStatus(fromBranch: $fromBranch, flakyRunsWindow: 50) {
+        __typename
+        ... on CloudProjectSpecFlakyStatus {
+          dashboardUrl
+        }
+      }
+    }
+  }
+}
+`
+
+const props = defineProps<{
+  projectGql: FlakyInformationProjectFragment | null | undefined
+  specGql: FlakyInformationSpecFragment | null | undefined
+  cloudSpecGql: FlakyInformationCloudSpecFragment | null | undefined
+}>()
+
+const isFlaky = computed(() => props.cloudSpecGql?.data?.__typename === 'CloudProjectSpec' && !!props.cloudSpecGql?.data?.isConsideredFlaky)
```

Comment:
Not really any error on your end but it drives me crazy just how many null checks we need (mainly due to GraphQL having a "all the things can be nullable" philosophy).

I know we can't really do much about it, but does anyone know if there's any other patterns? How do gigantic GraphQL apps with hundreds of fields handle this problem, do they all just write the code with lots of `?` fields?

Reply:
Ohhh I like the adapter pattern you are suggesting. The ugliness still exists, but you only need to look at it if you are actually working on it - right now, if you touch any UI component with graphql in a drive-by fashion, you have to see all this and grok what the heck is going on.

I'd like to see some experimentation with this, maybe in a non trivial and non time sensitive feature in the future.